Sources/JavaScriptKit/BasicObjects/JSPromise.swift

```diff
@@ -0,0 +1,136 @@
+/** A wrapper around [the JavaScript `Promise` class](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Promise)
+that exposes its functions in a type-safe and Swifty way. The `JSPromise` API is generic over both
+`Success` and `Failure` types, which improves compatibility with other statically-typed APIs such
+as Combine. If you don't know the exact type of your `Success` value, you should use `JSValue`, e.g.
+`JSPromise<JSValue, JSError>`. In the rare case, where you can't guarantee that the error thrown
+is of actual JavaScript `Error` type, you should use `JSPromise<JSValue, JSValue>`.
+
+This doesn't 100% match the JavaScript API, as `then` overload with two callbacks is not available.
+It's impossible to unify success and failure types from both callbacks in a single returned promise
+without type erasure. You should chain `then` and `catch` in those cases to avoid type erasure.
+*/
+public final class JSPromise<Success, Failure>: JSValueConvertible, JSValueConstructible
```

Comment:
Please note that this class must have the same lifetime with the actual `Promise` object in JS environment because callback handlers will be deallocated when `JSPromise.deinit`.

If the actual `Promise` object in JS environment lives longer than this `JSPromise`, it may call deallocated `JSClosure`.